### PR TITLE
Optimize GroupNorm on CPU

### PR DESCRIPTION
**Pull request by xiaomengy**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28204 Optimize GroupNorm on CUDA
* **#28203 Optimize GroupNorm on CPU**

Optimize GroupNorm on CPU.

Profiling result for input shape = [128, 256, 28, 28], num_groups = 32

Before this PR

```
GroupNorm forward: 210.97913278000487ms
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                        Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
native_batch_norm           68.00%           135.920s         68.00%           135.920s         135.920ms        NaN              0.000us          0.000us          1000             
addcmul                     31.94%           63.841s          31.94%           63.841s          63.841ms         NaN              0.000us          0.000us          1000             
view                        0.03%            69.572ms         0.03%            69.572ms         17.393us         NaN              0.000us          0.000us          4000             
group_norm                  0.01%            29.095ms         100.00%          199.870s         199.870ms        NaN              0.000us          0.000us          1000             
_batch_norm_impl_index      0.00%            5.179ms          68.01%           135.925s         135.925ms        NaN              0.000us          0.000us          1000             
batch_norm                  0.00%            4.594ms          68.01%           135.929s         135.929ms        NaN              0.000us          0.000us          1000             
contiguous                  0.00%            1.815ms          0.00%            1.815ms          1.815us          NaN              0.000us          0.000us          1000             
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 199.870s
CUDA time total: 0.000us

GroupNorm backward: 498.55486265799846ms
-----------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                                 Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
-----------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
mul                                  46.02%           224.071s         46.02%           224.071s         56.018ms         NaN              0.000us          0.000us          4000             
native_batch_norm_backward           43.86%           213.559s         43.86%           213.559s         213.559ms        NaN              0.000us          0.000us          1000             
sum                                  3.91%            19.040s          3.91%            19.040s          9.520ms          NaN              0.000us          0.000us          2000             
add_                                 3.28%            15.951s          3.28%            15.951s          5.317ms          NaN              0.000us          0.000us          3000             
AddcmulBackward                      1.95%            9.501s           47.97%           233.571s         233.571ms        NaN              0.000us          0.000us          1000             
torch::autograd::AccumulateGrad      0.97%            4.737s           4.25%            20.687s          6.896ms          NaN              0.000us          0.000us          3000             
as_strided                           0.00%            20.049ms         0.00%            20.049ms         5.012us          NaN              0.000us          0.000us          4000             
NativeBatchNormBackward              0.00%            13.951ms         43.86%           213.573s         213.573ms        NaN              0.000us          0.000us          1000             
reshape                              0.00%            13.034ms         0.01%            33.083ms         8.271us          NaN              0.000us          0.000us          4000             
ViewBackward                         0.00%            9.419ms          0.01%            42.502ms         10.626us         NaN              0.000us          0.000us          4000             
torch::autograd::GraphRoot           0.00%            1.189ms          0.00%            1.189ms          1.189us          NaN              0.000us          0.000us          1000             
-----------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 486.916s
CUDA time total: 0.000us
```

After this PR

```
GroupNorm forward: 59.529704386994126ms
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                   Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
native_group_norm      99.98%           56.154s          99.98%           56.154s          56.154ms         NaN              0.000us          0.000us          1000             
group_norm             0.02%            9.627ms          100.00%          56.164s          56.164ms         NaN              0.000us          0.000us          1000             
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 56.164s
CUDA time total: 0.000us

GroupNorm backward: 108.73398442400503ms
-------------------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                                               Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
-------------------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
non_differentiable_native_group_norm_backward      79.80%           84.432s          79.80%           84.432s          84.432ms         NaN              0.000us          0.000us          1000             
add_                                               15.72%           16.631s          15.72%           16.631s          5.544ms          NaN              0.000us          0.000us          3000             
torch::autograd::AccumulateGrad                    4.47%            4.725s           20.18%           21.356s          7.119ms          NaN              0.000us          0.000us          3000             
NativeGroupNormBackward                            0.01%            14.391ms         79.81%           84.446s          84.446ms         NaN              0.000us          0.000us          1000             
torch::autograd::GraphRoot                         0.00%            846.119us        0.00%            846.119us        0.846us          NaN              0.000us          0.000us          1000             
-------------------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 105.803s
CUDA time total: 0.000us
```

Differential Revision: [D17901506](https://our.internmc.facebook.com/intern/diff/D17901506/)

### CODE REVIEW

**Comment by BIT-silence** [comment]
link to https://github.com/pytorch/pytorch/issues/28201

**Comment by BIT-silence** [comment]
cc @ppwwyyxx

**Review by ngimel — CHANGES_REQUESTED** [review]
This generally looks good, but I have a suggestion regarding dispatch mechanism. You can either use device-specific dispatch defined in native_functions.yaml (there are many examples of that), or you can use stubs mechanism to achieve basically the same effect (see e.g. how it is done for indexing https://github.com/pytorch/pytorch/blob/15f9fe1d92a5d1e86278ae25f92dd9677b4956dc/aten/src/ATen/native/Normalization.cpp#L63-L71). Mixing the two is rarely if ever necessary, and creates a much more complicated code. Can you please stick to one of those? Otherwise the implementation itself is fine.

**Comment by bonlime** [comment]
Hi.
@BIT-silence are you planning to do the requested changes?

**Comment by BIT-silence** [comment]
Yes, just previously I don't have enough time to work on this. I will try to fix the things and merge it soon.

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  const T s = T(1) / static_cast<T>(D * HxW);
+  constexpr int64_t K = vec256::Vec256<T>::size();
+  const int64_t inner_size = D * HxW / K * K;
+  std::array<T, K> mean_arr;
+  std::array<T, K> rstd_arr;
+  const bool gamma_null = gamma_data == nullptr;
+  const bool beta_null = beta_data == nullptr;
+  for (int64_t i = 0; i < N; ++i) {
+    for (int64_t j = 0; j < G; ++j) {
+      const T* X_ptr = X_data + i * C * HxW + j * D * HxW;
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp]: Made it a single for-loop and also added multi-threading.

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    Tensor* dgamma,
+    Tensor* dbeta) {
+  DCHECK_EQ(dY.numel(), N * C * HxW);
+  DCHECK_EQ(X.numel(), N * C * HxW);
+  DCHECK_EQ(mean.numel(), N * group);
+  DCHECK_EQ(rstd.numel(), N * group);
+  DCHECK(!gamma.defined() || gamma.numel() == C);
+  const int64_t G = group;
+  const int64_t D = C / G;
+  const T* dY_data = dY.template data_ptr<T>();
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp]: Cleared.

**Comment by BIT-silence** [comment]
> This generally looks good, but I have a suggestion regarding dispatch mechanism. You can either use device-specific dispatch defined in native_functions.yaml (there are many examples of that), or you can use stubs mechanism to achieve basically the same effect (see e.g. how it is done for indexing
> 
> https://github.com/pytorch/pytorch/blob/15f9fe1d92a5d1e86278ae25f92dd9677b4956dc/aten/src/ATen/native/Normalization.cpp#L63-L71
> 
> ). Mixing the two is rarely if ever necessary, and creates a much more complicated code. Can you please stick to one of those? Otherwise the implementation itself is fine.

Done.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1316,6 +1316,16 @@
 
 - func: group_norm(Tensor input, int num_groups, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enabled=True) -> Tensor
 
+- func: native_group_norm(Tensor input, Tensor? weight, Tensor? bias, int N, int C, int HxW, int group, float eps) -> (Tensor, Tensor, Tensor)
+  dispatch:
```
Comment (ngimel) [inline aten/src/ATen/native/native_functions.yaml:1320]: do you need to define dispatch now? native_group_norm is not device-specific, it should work without dispatch.

**Review by ngimel — COMMENTED** [review]
Looks good, I have only minor comments.

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+    const Tensor& beta,
+    int64_t N,
+    int64_t C,
+    int64_t HxW,
+    int64_t group,
+    T eps,
+    Tensor* Y,
+    Tensor* mean,
+    Tensor* rstd) {
+  DCHECK_EQ(X.numel(), N * C * HxW);
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp]: Done.

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
Thanks!

**Pull request closed without merging**